rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp

```diff
@@ -191,25 +191,42 @@ void SequentialReader::set_filter(
 
 void SequentialReader::reset_filter()
 {
-  topics_filter_ = rosbag2_storage::StorageFilter();
+  set_filter(rosbag2_storage::StorageFilter());
+}
+
+void SequentialReader::seek(const rcutils_time_point_value_t & timestamp)
+{
+  seek_time_ = timestamp;
   if (storage_) {
-    storage_->reset_filter();
+    // reset to the first file
+    current_file_iterator_ = file_paths_.begin();
+    load_current_file();
```

Comment:
We are resetting to the first file and trying to `seek` only in it. But what if requested timestamp in second or third file?
It seems we are not going to try to search out of the first file. 
Am I missed some logic?

Reply:
When we `read_next()`, we call `has_next()`, which will call `load_next_file()` if we reach the end of the file and there is another file. It will do so until either there is a next message found, or until there are no more files to open. Resetting to the first file guarantees that you will be able to roll through to subsequent files while reading. Not resetting to the first file means that you will miss messages that occurred before the current file.

Example:

```
##### File 0 #####
t0 msg0
t1 msg1
t2 msg2

##### File 1 #####
t3 msg3
t4 msg4
t5 msg5
```

Assume we just finished reading `msg3`, and the next message is `msg4`. When we call `seek(t1)`, it will set the current file to `File 0`, and open `File 0` and set filter and seek time. Then subsequent calls to `read_next()` will return:

```
msg1
msg2
msg3  <-- rollover to File 1 happens on the has_next() call in this read_next() call
msg4
msg5
```

If we don't reset the current file to the first file, then instead, we will only be seeking within the current file before the `seek(t1)` call and our reads will return:

```
msg3  <-- First msg in File 1 to occur after t1
msg4
msg5
```

Now consider the case of calling `seek(t5)`. If we reset to the first file (File 0), then on the first `read_next()`, we will be calling `has_next()` which calls `storage_->has_next()`, which will return `false`. This will trigger the reader to check for and load the next file (File 1), and we proceed to seeking within File 1:

```
// Finds no messages in File 0, rolls over to File 1
msg5 <-- First message in File 1 that occurs on or after t5
```

Reply:
btw the new test cases in rosbag2_py is essentially testing that `seek(t)` behaves as expected backward, forward and across file splits. You're welcome print out the entirety of the `wbag` messages and follow along.

Reply:
@lihui815 Thank you for detailed explanation. I missed the part that we start calling `has_next` inside `read_next` API recently.